runner/runner.py

```diff
@@ -19,46 +18,52 @@
 obligatory_plugins = [PipelineAnalyser(), IntegrityChecker()]
 
 
-def overwrite_model_configs(config: RunConfig, pipeline: Pipeline) -> Pipeline:
+def overwrite_model_configs_(config: RunConfig, pipeline: Pipeline) -> None:
     for key, value in vars(config).items():
         if value is not None:
             for model in flatten(pipeline.children()):
                 if hasattr(model, "config"):
                     if hasattr(model.config, key):
                         vars(model.config)[key] = value
 
-    return pipeline
-
-
-def add_position_to_block_names(pipeline: Pipeline) -> Pipeline:
-    entire_pipeline = pipeline.children()
-
-    def add_position(block: Union[List[Element], Element], position: int, prefix: str):
-        if isinstance(block, List):
-            if position > 0:
-                prefix += f"{position - 1}-"
-            for i, child in enumerate(block):
-                add_position(child, i, prefix)
-        elif not isinstance(block, DataSource):
-            block.id += f"{prefix}{position}"
-
-    add_position(entire_pipeline, 1, "-")
-
-    return pipeline
-
 
-def append_pipeline_id(pipeline: Pipeline) -> Pipeline:
+def append_parent_path_and_id(pipeline: Pipeline) -> None:
     entire_pipeline = pipeline.dict_children()
 
-    def append_id(block, pipeline_id: str):
-        block["obj"].pipeline_id = f"{pipeline_id}"
+    blocks_encountered = []
 
-        if "children" in block:
-            for child in block["children"]:
-                append_id(child, f"{pipeline_id}/{block['name']}")
+    def append(block, parent_path: str, id_with_prefix: str):
+        block["obj"].parent_path = f"{parent_path}"
+        block["obj"].id += f"{id_with_prefix}"
+        blocks_encountered.append(id(block["obj"]))
 
-    append_id(entire_pipeline, Const.output_pipelines_path)
-    return pipeline
+        if "children" in block:
+            for i, child in enumerate(block["children"]):
+                child_id = (
+                    f"{id_with_prefix}-{i}"
+                    if id(child["obj"]) not in blocks_encountered
```

Comment:
question: can we rely on `id()` here, now that we know that deepcopy doesn't change the id?

Reply:
This is local tracking of the pipeline blocks so it simply checks if there is a duplicate or not. ``deepcopy`` happens before it so it shouldn't interfere with this loop.